### PR TITLE
pinned events: limit the max amount of concurrent requests to load pinned events

### PR DESCRIPTION
**Pull request by jmartinesp**
Depends on https://github.com/matrix-org/matrix-rust-sdk/pull/3840.

`RequestConfig::max_concurrent_requests` won't work for requests to a same endpoint, since it's a global value that can only be passed to `HttpClient` when created.

To replace it I added a buffered stream that can run at most `MAX_PINNED_EVENTS_CONCURRENT_REQUESTS` (10) requests concurrently.

There is also a test to make sure this max concurrency mechanism works.

- [ ] Public API changes documented in changelogs (optional)

Signed-off-by:

### CODE REVIEW

File: crates/matrix-sdk-ui/tests/integration/timeline/pinned_event.rs
Comment (andybalaam) [inline crates/matrix-sdk-ui/tests/integration/timeline/pinned_event.rs:699]: Would it be safer to sleep for `1.5 * request_delay` just to make sure some other delay doesn't cause us to just miss it?

**Review by andybalaam — APPROVED** [review]
Looks good, one question.

File: crates/matrix-sdk-ui/tests/integration/timeline/pinned_event.rs
Comment (andybalaam) [reply crates/matrix-sdk-ui/tests/integration/timeline/pinned_event.rs:699]: Then `0.5 * request_delay`?

Comment (jmartinesp) [reply crates/matrix-sdk-ui/tests/integration/timeline/pinned_event.rs:699]: Ok, sounds good. I wanted to use the full delay to prove this worked as expected, but it's true it might result in flaky tests 🫤 .

File: crates/matrix-sdk-ui/src/timeline/pinned_events_loader.rs
Comment (Hywan) [inline crates/matrix-sdk-ui/src/timeline/pinned_events_loader.rs:29]: If it's a `pub` item, please add a documentation to it.

If it's a `pub` only for testing purposes, maybe adding a `#[cfg(test)]` would be nice?

**Review by Hywan — CHANGES_REQUESTED** [review]

File: crates/matrix-sdk-ui/src/timeline/pinned_events_loader.rs
Comment (jmartinesp) [reply crates/matrix-sdk-ui/src/timeline/pinned_events_loader.rs:29]: > If it's a pub item, please add a documentation to it.

True, my bad.

> If it's a pub only for testing purposes, maybe adding a #[cfg(test)] would be nice?

Well, it just needs to be `pub` for the tests, but it needs to exist in general since it's used in the code below it. I don't know if there's some way to do that or it's better to just keep it public.

Comment (Hywan) [reply crates/matrix-sdk-ui/src/timeline/pinned_events_loader.rs:30]: That's one limitation of Rust where we can add a `#[cfg]` for a visiblity only. There is a crate to handle that, https://crates.io/crates/cfg-vis, but [we have removed it recently](https://github.com/matrix-org/matrix-rust-sdk/pull/3375).

You can write this as follows:

```rust
#[cfg(test)]
pub const … = …;

#[cfg(not(test))]
const … = …;
```

if you don't want to repeat the constant value, you can have a private constant just for that:

```rust
const VALUE: usize = 10;

#[cfg(test)]
pub const …: usize = VALUE;

#[cfg(not(test))]
const …: usize = VALUE;
```

of course, with better names that `…` and `VALUE` :-D.

Comment (jmartinesp) [reply crates/matrix-sdk-ui/src/timeline/pinned_events_loader.rs:30]: It seems like this won't work as the target test is an integration one.

File: crates/matrix-sdk-ui/src/timeline/pinned_events_loader.rs
Comment (bnjbvr) [inline crates/matrix-sdk-ui/src/timeline/pinned_events_loader.rs:74]: Wonder if we could use a plain Semaphore instead of a stream here, it's a bit overkill to use the stream machinery for this IMO…

Comment (jmartinesp) [reply crates/matrix-sdk-ui/src/timeline/pinned_events_loader.rs:74]: We could and did, but the code is more readable like this (IMO) and there's no performance penalty as far as I could tell by the results of the benchmark. Also, it's used in a very similar way [in some other parts of the SDK](https://github.com/matrix-org/matrix-rust-sdk/blob/224292ab3e3425600a6ee25aa5df17eb197c3831/crates/matrix-sdk/src/encryption/mod.rs#L723).

If we wanted to use a `Semaphore` instead, what would the `semaphore.acquire()` code look like if the result contains an error? I remember we discussed it in a previous PR but I'm not sure if we ever make a decision.

Comment (bnjbvr) [reply crates/matrix-sdk-ui/src/timeline/pinned_events_loader.rs:74]: > the code is more readable like this (IMO)

I find the code much worse, because it involves other async concepts that one must be aware of (streams), and higher-level typing that's hard to grasp just from reading the code (stream flattening). But let's keep it like this now, and we can change that later.

File: crates/matrix-sdk-ui/src/timeline/pinned_events_loader.rs
Comment (jmartinesp) [reply crates/matrix-sdk-ui/src/timeline/pinned_events_loader.rs:31]: Can I have another review, pretty please 🙏 ?

**Comment by bnjbvr** [comment]
No, PRs need to be approved before getting merged.

**Comment by jmartinesp** [comment]
Then we're lucky we're not in a rush, I guess.

File: crates/matrix-sdk-ui/src/timeline/pinned_events_loader.rs
Comment (bnjbvr) [reply crates/matrix-sdk-ui/src/timeline/pinned_events_loader.rs:29]: > It seems like this won't work as the target test is an integration one.

Why? Does `#[cfg(any(feature = "testing", test))]` not work?

Comment (jmartinesp) [reply crates/matrix-sdk-ui/src/timeline/pinned_events_loader.rs:29]: I didn't try that exact configuration, let me try with that and report back.

Comment (jmartinesp) [reply crates/matrix-sdk-ui/src/timeline/pinned_events_loader.rs:29]: It doesn't seem to work:

```rust
#[cfg(any(feature = "testing", test))]
pub const MAX_CONCURRENT_REQUESTS: usize = 10;

... // then

#[cfg(any(feature = "testing", test))]
pub use self::pinned_events_loader::MAX_CONCURRENT_REQUESTS;
```

Ends up with this error, I tried several variations:

```
error[E0432]: unresolved import `matrix_sdk_ui::timeline::MAX_CONCURRENT_REQUESTS`
   --> crates/matrix-sdk-ui/tests/integration/timeline/pinned_event.rs:35:5
    |
35  | use matrix_sdk_ui::timeline::MAX_CONCURRENT_REQUESTS;
    |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ no `MAX_CONCURRENT_REQUESTS` in `timeline`
    |
note: found an item that was configured out
   --> /Users/jorge/Developer/Element/matrix-rust-sdk/crates/matrix-sdk-ui/src/timeline/mod.rs:110:37
    |
110 | pub use self::pinned_events_loader::MAX_CONCURRENT_REQUESTS;
    |                                     ^^^^^^^^^^^^^^^^^^^^^^^
```

Comment (bnjbvr) [reply crates/matrix-sdk-ui/src/timeline/pinned_events_loader.rs:29]: looks like you need a bit more cfg guards then :thinking:

Comment (jmartinesp) [reply crates/matrix-sdk-ui/src/timeline/pinned_events_loader.rs:29]: Any idea what those might be? Are the tests in `tests/integration` running in a different way than the rest?

Comment (jmartinesp) [reply crates/matrix-sdk-ui/src/timeline/pinned_events_loader.rs:29]: From the [Rust docs](https://doc.rust-lang.org/book/ch11-03-test-organization.html#integration-tests):

> In Rust, integration tests are entirely external to your library. They use your library in the same way any other code would, which means they can only call functions that are part of your library’s public API. Their purpose is to test whether many parts of your library work together correctly. Units of code that work correctly on their own could have problems when integrated, so test coverage of the integrated code is important as well. To create integration tests, you first need a tests directory.

So I guess the test and testing cfgs are not used and we're back to "It seems like this won't work as the target test is an integration one." 🫤 .

Comment (jmartinesp) [reply crates/matrix-sdk-ui/src/timeline/pinned_events_loader.rs:29]: Since this apparaently can't be done and the doc comment was added as @Hywan requested, can we have another review so we can at least make some progress?

Comment (bnjbvr) [reply crates/matrix-sdk-ui/src/timeline/pinned_events_loader.rs:29]: I've found a better idea, but it involves a bit more work (sowwy). Since the `PinnedEventLoader` can use an arbitrary `PinnedEventsRoom` impl, can you make the test a non-integration test, at the bottom of this file, using a test impl for `PinnedEventsRoom`? Maybe running the assertions will be even simpler with that, as we don't have to rely on wiremock anymore?

Comment (jmartinesp) [reply crates/matrix-sdk-ui/src/timeline/pinned_events_loader.rs:29]: Honestly, wiremock is quite the help here... I'd rather not change everything for just a `pub` modifier. Maybe I can add this to be part of the `TimelineFocus::PinnedEvents` instead, like `max_events_to_load`?

Comment (bnjbvr) [reply crates/matrix-sdk-ui/src/timeline/pinned_events_loader.rs:29]: It would also make the testing more targeted by not involved network, and make it much faster to run, but I understand this is a big ask.

> Maybe I can add this to be part of the TimelineFocus::PinnedEvents instead, like max_events_to_load?

Yep, sounds great, thanks!

File: crates/matrix-sdk-ui/src/timeline/pinned_events_loader.rs
Comment (bnjbvr) [inline crates/matrix-sdk-ui/src/timeline/pinned_events_loader.rs:75]: Can you inline this variable, as it's only used once?

File: crates/matrix-sdk-ui/tests/integration/timeline/pinned_event.rs
Comment (bnjbvr) [inline crates/matrix-sdk-ui/tests/integration/timeline/pinned_event.rs:145]: Can you make a tiny test helper that creates a pinned events focused timeline, instead? that will avoid repeating the new argument everywhere

File: crates/matrix-sdk-ui/tests/integration/timeline/pinned_event.rs
Comment (bnjbvr) [inline crates/matrix-sdk-ui/tests/integration/timeline/pinned_event.rs:705]: Is 50 milliseconds enough for the code cov task? That would be fine to use something greater up to 1 second, fwiw.

**Review by bnjbvr — APPROVED** [review]
Thanks, approving assuming my comments will be addressed.

File: crates/matrix-sdk-ui/tests/integration/timeline/pinned_event.rs
Comment (jmartinesp) [reply crates/matrix-sdk-ui/tests/integration/timeline/pinned_event.rs:705]: From the codecov report it seems like those lines are counted as covered too, so maybe there's no need to increase this.

**Review by Hywan — APPROVED** [review]

**Pull request merged**
